Modules/_abc.c

```diff
@@ -0,0 +1,1049 @@
+    if (!result) {
+        Py_DECREF(subclass);
+        return NULL;
+    }
+    if (result == Py_True) {
+        Py_DECREF(subclass);
+        return Py_True;
+    }
+    Py_DECREF(result);
+    Py_DECREF(subclass);
+    return PyObject_CallMethod(self, "__subclasscheck__", "O", subtype);
+}
+
+
+PyDoc_STRVAR(_abc_subclasscheck_doc,
+"Internal ABC helper for subclasss checks. Should be never used outside abc module");
+
+static PyObject *
+_abc_subclasscheck(PyObject *m, PyObject *args)
+{
+    PyObject *self, *subclasses, *subclass = NULL;
+    PyObject *ok, *mro, *key, *rkey;
+    Py_ssize_t pos;
+    int incache, result;
+    if (!PyArg_UnpackTuple(args, "_abc_subclasscheck", 2, 2, &self, &subclass)) {
+        return NULL;
+    }
+
+    /* 1. Check cache. */
+    incache = _in_cache(self, subclass);
+    if (incache < 0) {
+        return NULL;
+    }
+    if (incache > 0) {
+        Py_RETURN_TRUE;
+    }
+    /* 2. Check negative cache; may have to invalidate. */
+    incache = _in_negative_cache(self, subclass);
+    if (incache < 0) {
+        return NULL;
+    }
+    if (PyObject_RichCompareBool(_get_negative_cache_version(self),
+                                 abc_invalidation_counter, Py_LT) > 0) {
+        /* Invalidate the negative cache. */
+        if (_reset_negative_cache(self) < 0) {
+            return NULL;
+        }
+        if (_set_negative_cache_version(self, abc_invalidation_counter) < 0) {
+            return NULL;
+        }
+    } else if (incache) {
+        Py_RETURN_FALSE;
+    }
+    /* 3. Check the subclass hook. */
+    ok = PyObject_CallMethod((PyObject *)self, "__subclasshook__", "O", subclass);
+    if (!ok) {
+        return NULL;
+    }
+    if (ok == Py_True) {
+        if (_add_to_cache(self, subclass) < 0) {
+            Py_DECREF(ok);
+            return NULL;
+        }
+        return Py_True;
+    }
+    if (ok == Py_False) {
+        if (_add_to_negative_cache(self, subclass) < 0) {
+            Py_DECREF(ok);
+            return NULL;
+        }
+        return Py_False;
+    }
+    if (ok != Py_NotImplemented) {
+        Py_DECREF(ok);
+        PyErr_SetString(PyExc_AssertionError, "__subclasshook__ must return either"
+                                              " False, True, or NotImplemented");
+        return NULL;
+    }
+    Py_DECREF(ok);
+    /* 4. Check if it's a direct subclass. */
+    mro = ((PyTypeObject *)subclass)->tp_mro;
+    for (pos = 0; pos < PyTuple_Size(mro); pos++) {
+        if ((PyObject *)self == PyTuple_GetItem(mro, pos)) {
+            if (_add_to_cache(self, subclass) < 0) {
+                return NULL;
+            }
+            Py_RETURN_TRUE;
+        }
+    }
+
+    /* 5. Check if it's a subclass of a registered class (recursive). */
+    pos = 0;
+    Py_hash_t hash;
+    if (_enter_iter(self) < 0) {
+        return NULL;
+    }
+    while (_PySet_NextEntry(_get_registry(self), &pos, &key, &hash)) {
```

Comment:
The result of `_get_registry` needs to be checked explicitly before calling `_PySet_NextEntry` to handle the error case.